chapter_recurrent-neural-networks/sequence.ipynb

```diff
@@ -0,0 +1,584 @@
+    "\n",
+    "for (int i = 0; i < tau; i++) {\n",
+    "    features.set(new NDIndex(\":, \" + i), x.get(new NDIndex(\"\" + i + \":\" + (T - tau + i) )));\n",
+    "}\n",
+    "NDArray labels = x.get(new NDIndex(\"\" + tau + \":\")).reshape(new Shape(-1,1));"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "int batchSize = 16;\n",
+    "int nTrain = 600;\n",
+    "// Only the first `nTrain` examples are used for training\n",
+    "ArrayDataset trainIter = new ArrayDataset.Builder()\n",
+    "    .setData(features.get(new NDIndex(\":\" + nTrain)))\n",
+    "    .optLabels(labels.get(new NDIndex(\":\" + nTrain)))\n",
+    "    .setSampling(batchSize, false)\n",
+    "    .build()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 10
+   },
+   "source": [
+    "Here we keep the architecture fairly simple:\n",
+    "just an MLP with two fully-connected layers, ReLU activation, and squared loss.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "// A simple MLP\n",
+    "public SequentialBlock getNet() {\n",
+    "    SequentialBlock net = new SequentialBlock();\n",
+    "//     net.add(Blocks.batchFlattenBlock(4));\n",
+    "    net.add(Linear.builder().setUnits(10).build());\n",
+    "    net.add(Activation::relu);\n",
+    "    net.add(Linear.builder().setUnits(1).build());\n",
+    "    \n",
+    "    net.setInitializer(new XavierInitializer());\n",
+    "    net.initialize(manager, DataType.FLOAT32, new Shape(4));\n",
+    "    return net;\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 14
+   },
+   "source": [
+    "Now we are ready to train the model. The code below is essentially identical to the training loop in previous sections,\n",
+    "such as :numref:`sec_linear_concise`.\n",
+    "Thus, we will not delve into much detail.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "public Model train(SequentialBlock net, ArrayDataset dataset, int batchSize, int numEpochs, float learningRate) {\n",
+    "    // Square Loss\n",
+    "    Loss loss = Loss.l2Loss();\n",
+    "    Tracker lrt = Tracker.fixed(learningRate);\n",
+    "    Optimizer sgd = Optimizer.adam().optLearningRateTracker(lrt).build();\n",
+    "    \n",
+    "    DefaultTrainingConfig config = new DefaultTrainingConfig(loss)\n",
+    "        .optOptimizer(sgd) // Optimizer (loss function)\n",
+    "        .addTrainingListeners(TrainingListener.Defaults.logging()); // Logging\n",
+    "    \n",
+    "    Model model = Model.newInstance(\"sequence\");\n",
+    "    model.setBlock(net);\n",
+    "    Trainer trainer = model.newTrainer(config);\n",
+    "    \n",
+    "    trainer.initialize(new Shape(batchSize,4));\n",
```

Comment:
you can also remove initialize here, we run initialize under the hood if you don't

Reply:
Done